### PR TITLE
Verify transactions are well formed when computing the block's data hash instead of externally

### PR DESCRIPTION
**Pull request by yacovm**
To make sure the function VerifyTransactionsAreWellFormed is called in the right places,
I moved its invocation to the computation of the Fabric's block data hash.

### CODE REVIEW

File: orderer/consensus/smartbft/assembler.go [hunk 1]
```diff
@@ -51,7 +53,11 @@ func (a *Assembler) AssembleProposal(metadata []byte, requests [][]byte) (nextPr
 
 	block := protoutil.NewBlock(lastBlock.Header.Number+1, protoutil.BlockHeaderHash(lastBlock.Header))
 	block.Data = &cb.BlockData{Data: batchedRequests}
-	block.Header.DataHash = protoutil.BlockDataHash(block.Data)
+
+	// Instead of calling BlockDataHash() which calls VerifyTransactionsAreWellFormed(),
+	// we directly compute the block's data hash.
+	dataHash := sha256.Sum256(bytes.Join(block.Data.Data, nil))
+	block.Header.DataHash = dataHash[:]
```
Comment (tock-ibm) [inline orderer/consensus/smartbft/assembler.go]: I don't like this... I think the mix of calling BlockDataHash and direct computation like here is s recipe for trouble. Especially if we go for an update of said computation in V3. 

It would be much better in my opinion to offer two  versions of the original BlockDataHash function, one that does verify and one that does not. 
You can change the name / signature and this way we'll catch all the instances. For example:

BlockDataHash(...) ([]byte, error) \\ with verify

BlockDataHashCompute (...) ([]byte) \\ w/o verify

or maybe add a bool to the BlockDataHash (bc, verfyTxs bool) that selects whether to verify or not.

or finally, bite the bullet and verify even if you don't need to.

File: common/ledger/blockledger/util.go [hunk 2]
```diff
@@ -76,7 +76,7 @@ func CreateNextBlock(rl Reader, messages []*cb.Envelope) *cb.Block {
 	}
 
 	block := protoutil.NewBlock(nextBlockNumber, previousBlockHash)
-	block.Header.DataHash = protoutil.BlockDataHash(data)
+	block.Header.DataHash, _ = protoutil.BlockDataHash(data)
```
Comment (tock-ibm) [inline common/ledger/blockledger/util.go]: Here you use the version that verifies the TXs even though it is really unnecessary. In another place below you use the raw computation... I think a uniform approach is better. See comment bellow.

**Comment by tock-ibm** [comment]
I have to admit that I am not entirely comfortable with this approach. There are a lot of places where the TXs verification is not necessary, and so the error is just ignored... so why not replace all the original instances of BlockDataHash with two versions: one that does not verify, and one that does? reviewing such a change is exactly like reviewing this one, and having to decide whether it is OK not to check the error or it is a must. Plus, it avoids checking the TXs in places where it is really not needed.

**Comment by yacovm** [comment]
> I have to admit that I am not entirely comfortable with this approach. There are a lot of places where the TXs verification is not necessary, and so the error is just ignored... so why not replace all the original instances of BlockDataHash with two versions: one that does not verify, and one that does? reviewing such a change is exactly like reviewing this one, and having to decide whether it is OK not to check the error or it is a must. Plus, it avoids checking the TXs in places where it is really not needed.

Done.

File: cmd/osnadmin/main_test.go [hunk 3]
```diff
@@ -820,7 +820,7 @@ func blockWithGroups(groups map[string]*cb.ConfigGroup, channelID string) *cb.Bl
 			}),
 		},
 	}
-	block.Header.DataHash = protoutil.BlockDataHash(block.Data)
+	block.Header.DataHash, _ = protoutil.BlockDataHash(block.Data)
```
Comment (tock-ibm) [inline cmd/osnadmin/main_test.go:823]: Why not ComputeBlockDataHash?

File: common/ledger/testutil/test_helper.go [hunk 4]
```diff
@@ -303,7 +307,7 @@ func NewBlock(env []*common.Envelope, blockNum uint64, previousHash []byte) *com
 		txEnvBytes, _ := proto.Marshal(env[i])
 		block.Data.Data = append(block.Data.Data, txEnvBytes)
 	}
-	block.Header.DataHash = protoutil.BlockDataHash(block.Data)
+	block.Header.DataHash, _ = protoutil.BlockDataHash(block.Data)
```
Comment (tock-ibm) [inline common/ledger/testutil/test_helper.go]: Why not ComputeBlockDataHash?

File: core/committer/txvalidator/v14/txvalidator_test.go [hunk 5]
```diff
@@ -277,9 +277,10 @@ func TestTxValidationFailure_InvalidTxid(t *testing.T) {
 		},
 	}
 
+	hash, _ := protoutil.BlockDataHash(block.Data)
```
Comment (tock-ibm) [inline core/committer/txvalidator/v14/txvalidator_test.go:280]: Why not ComputeBlockDataHash? or check the error?

File: core/committer/txvalidator/v20/txvalidator_test.go [hunk 6]
```diff
@@ -231,9 +231,10 @@ func TestTxValidationFailure_InvalidTxid(t *testing.T) {
 		},
 	}
 
+	hash, _ := protoutil.BlockDataHash(block.Data)
```
Comment (tock-ibm) [inline core/committer/txvalidator/v20/txvalidator_test.go]: Why not ComputeBlockDataHash? or check the error?

File: core/ledger/kvledger/benchmark/chainmgmt/block_gen.go [hunk 7]
```diff
@@ -82,7 +82,7 @@ func (bg *blkGenerator) nextBlock() *common.Block {
 	if len(block.Data.Data) == 0 {
 		return nil
 	}
-	block.Header.DataHash = protoutil.BlockDataHash(block.Data)
+	block.Header.DataHash, _ = protoutil.BlockDataHash(block.Data)
```
Comment (tock-ibm) [inline core/ledger/kvledger/benchmark/chainmgmt/block_gen.go]: Why not ComputeBlockDataHash?

File: core/ledger/kvledger/tests/committer.go [hunk 8]
```diff
@@ -90,7 +90,7 @@ func (g *blkGenerator) nextBlockAndPvtdata(trans []*txAndPvtdata, missingPvtData
 			blockPvtdata[seq] = &ledger.TxPvtData{SeqInBlock: seq, WriteSet: tran.Pvtws}
 		}
 	}
-	block.Header.DataHash = protoutil.BlockDataHash(block.Data)
+	block.Header.DataHash, _ = protoutil.BlockDataHash(block.Data)
```
Comment (tock-ibm) [inline core/ledger/kvledger/tests/committer.go]: Why not ComputeBlockDataHash?

File: integration/raft/channel_participation_test.go [hunk 9]
```diff
@@ -1081,7 +1081,7 @@ func createJoinBlockDefineSystemChannel(channelID string) *common.Block {
 			}),
 		},
 	}
-	block.Header.DataHash = protoutil.BlockDataHash(block.Data)
+	block.Header.DataHash, _ = protoutil.BlockDataHash(block.Data)
```
Comment (tock-ibm) [inline integration/raft/channel_participation_test.go]: Why not ComputeBlockDataHash?

File: integration/raft/config_test.go [hunk 10]
```diff
@@ -188,7 +188,7 @@ var _ = Describe("EndToEnd reconfiguration and onboarding", func() {
 			Expect(err).NotTo(HaveOccurred())
 			genesisBlock.Data.Data[0], err = protoutil.Marshal(envelope)
 			Expect(err).NotTo(HaveOccurred())
-			genesisBlock.Header.DataHash = protoutil.BlockDataHash(genesisBlock.Data)
+			genesisBlock.Header.DataHash, _ = protoutil.BlockDataHash(genesisBlock.Data)
```
Comment (tock-ibm) [inline integration/raft/config_test.go]: Why not ComputeBlockDataHash?

File: integration/raft/config_test.go [hunk 11]
```diff
@@ -2127,7 +2127,7 @@ func configToGenesisBlock(configBlock *common.Block, channelId string) *common.B
 	genesisBlock.Data.Data[0], err = protoutil.Marshal(envelope)
 	Expect(err).NotTo(HaveOccurred())
 
-	genesisBlock.Header.DataHash = protoutil.BlockDataHash(genesisBlock.Data)
+	genesisBlock.Header.DataHash, _ = protoutil.BlockDataHash(genesisBlock.Data)
```
Comment (tock-ibm) [inline integration/raft/config_test.go]: Why not ComputeBlockDataHash?

File: integration/smartbft/smartbft_test.go [hunk 12]
```diff
@@ -1988,7 +1989,7 @@ func createPrePrepareRequest(
 			protoutil.MarshalOrPanic(env),
 		},
 	}
-	block.Header.DataHash = protoutil.BlockDataHash(block.Data)
+	block.Header.DataHash, _ = protoutil.BlockDataHash(block.Data)
```
Comment (tock-ibm) [inline integration/smartbft/smartbft_test.go]: Why not ComputeBlockDataHash?

File: internal/ledgerutil/verify/verify.go [hunk 13]
```diff
@@ -229,8 +229,12 @@ func checkBlock(block *common.Block, previousHash []byte) (*blockCheckResult, er
 		Errors:   []string{},
 	}
 
+	hash, err := protoutil.BlockDataHash(block.Data)
+	if err != nil {
+		return nil, err
+	}
```
Comment (tock-ibm) [inline internal/ledgerutil/verify/verify.go]: Why not add to the results? this is a utility to verify existing ledgers, and discovering such an error should be reported in the tool output, not crash the tool.

Comment (tock-ibm) [reply internal/ledgerutil/verify/verify.go]: In addition, a test case with such an error should be added to the unit test of 
`func TestVerify(t *testing.T)`  of verify_test.go

File: internal/peer/gossip/mcs_test.go [hunk 14]
```diff
@@ -475,7 +475,7 @@ func mockBlock(t *testing.T, channel string, seqNum uint64, localSigner *mocks.S
 	if len(dataHash) != 0 {
 		block.Header.DataHash = dataHash
 	} else {
-		block.Header.DataHash = protoutil.BlockDataHash(block.Data)
+		block.Header.DataHash, _ = protoutil.BlockDataHash(block.Data)
```
Comment (tock-ibm) [inline internal/peer/gossip/mcs_test.go]: why not ComputeBlockDataHash?

File: internal/peer/gossip/mcs_test.go [hunk 15]
```diff
@@ -518,7 +518,7 @@ func mockBlockBFT(t *testing.T, channel string, seqNum uint64, localSigner *mock
...
```
Comment (tock-ibm) [inline internal/peer/gossip/mcs_test.go]: why not ComputeBlockDataHash?

File: orderer/common/channelparticipation/validator.go [hunk 16]
```diff
@@ -31,7 +31,11 @@ func ValidateJoinBlock(configBlock *cb.Block) (channelID string, err error) {
 		return "", errors.New("invalid block: does not have metadata")
 	}
 
-	if !bytes.Equal(protoutil.BlockDataHash(configBlock.Data), configBlock.Header.DataHash) {
+	dataHash, err := protoutil.BlockDataHash(configBlock.Data)
+	if err != nil {
+		return "", err
+	}
```
Comment (tock-ibm) [inline orderer/common/channelparticipation/validator.go]: Add a unit test for when that happens, please

File: orderer/common/channelparticipation/validator_test.go [hunk 17]
```diff
@@ -162,7 +162,7 @@ func blockWithGroups(groups map[string]*cb.ConfigGroup, channelID string) *cb.Bl
 			}),
 		},
 	}
-	block.Header.DataHash = protoutil.BlockDataHash(block.Data)
+	block.Header.DataHash, _ = protoutil.BlockDataHash(block.Data)
```
Comment (tock-ibm) [inline orderer/common/channelparticipation/validator_test.go]: why not ComputeBlockDataHash?

File: orderer/common/channelparticipation/validator_test.go [hunk 18]
```diff
@@ -183,7 +183,7 @@ func nonConfigBlock() *cb.Block {
...
```
Comment (tock-ibm) [inline orderer/common/channelparticipation/validator_test.go:186]: why not ComputeBlockDataHash?

File: orderer/common/cluster/util.go [hunk 19]
```diff
@@ -260,11 +260,11 @@ func VerifyBlockHash(indexInBuffer int, blockBuff []*common.Block) error {
 	if block.Header == nil {
 		return errors.New("missing block header")
 	}
-	if err := protoutil.VerifyTransactionsAreWellFormed(block); err != nil {
+	seq := block.Header.Number
+	dataHash, err := protoutil.BlockDataHash(block.Data)
+	if err != nil {
```
Comment (tock-ibm) [inline orderer/common/cluster/util.go]: As mentioned in the previous PR, this will crash when `block.Data==nil`, see comment below.

File: protoutil/blockutils.go [hunk 20]
```diff
@@ -300,18 +307,18 @@ func searchConsenterIdentityByID(consenters []*cb.Consenter, identifier uint32)
 	return nil
 }
 
-func VerifyTransactionsAreWellFormed(block *cb.Block) error {
-	if block == nil || block.Data == nil || len(block.Data.Data) == 0 {
+func VerifyTransactionsAreWellFormed(bd *cb.BlockData) error {
+	if bd.Data == nil || len(bd.Data) == 0 {
```
Comment (tock-ibm) [inline protoutil/blockutils.go]: you also have to check against `bd == nil`

File: protoutil/blockutils_test.go [hunk 21]
```diff
@@ -36,15 +38,17 @@ func TestNewBlock(t *testing.T) {
 	require.Equal(t, []byte("datahash"), block.Header.PreviousHash, "Incorrect previous hash")
 	require.NotNil(t, block.GetData())
 	require.NotNil(t, block.GetMetadata())
-	block.GetHeader().DataHash = protoutil.BlockDataHash(data)
+	block.GetHeader().DataHash, _ = protoutil.BlockDataHash(data)
+
+	dataHash, _ := protoutil.BlockDataHash(data)
```
Comment (tock-ibm) [inline protoutil/blockutils_test.go]: check no error?
Is there a test case that returns error?
Check when `data` is nil?

File: protoutil/blockutils_test.go [hunk 22]
```diff
@@ -606,11 +610,15 @@ func TestVerifyTransactionsAreWellFormed(t *testing.T) {
 		},
 	} {
 		t.Run(tst.name, func(t *testing.T) {
-			err := protoutil.VerifyTransactionsAreWellFormed(tst.block)
-			if tst.expectedError == "" {
-				require.NoError(t, err)
+			if tst.block == nil || tst.block.Data == nil {
+				require.Error(t, errors.New("empty block"))
```
Comment (tock-ibm) [inline protoutil/blockutils_test.go]: what does this check? that an error is an error ;-)

File: orderer/common/cluster/util_test.go [hunk 23]
```diff
@@ -194,7 +194,7 @@ func TestVerifyBlockBFT(t *testing.T) {
 	configTx.Signature = []byte{1, 2, 3}
 	block.Data.Data[0] = protoutil.MarshalOrPanic(configTx)
 	block.Header.Number = 2
-	block.Header.DataHash = protoutil.BlockDataHash(block.Data)
+	block.Header.DataHash, _ = protoutil.BlockDataHash(block.Data)
```
Comment (tock-ibm) [inline orderer/common/cluster/util_test.go]: why not ComputeBlockDataHash?

File: orderer/common/cluster/util_test.go [hunk 24]
```diff
@@ -334,7 +334,7 @@ func createBlockChain(start, end uint64) []*common.Block {
 	var blockchain []*common.Block
 	for seq := start; seq <= end; seq++ {
 		block := newBlock(seq)
-		block.Header.DataHash = protoutil.BlockDataHash(block.Data)
+		block.Header.DataHash, _ = protoutil.BlockDataHash(block.Data)
```
Comment (tock-ibm) [inline orderer/common/cluster/util_test.go]: why not ComputeBlockDataHash?

File: orderer/common/follower/follower_chain_test.go [hunk 25]
```diff
@@ -945,11 +945,11 @@ func (mbc *memoryBlockChain) fill(numBlocks uint64) {
 		var block *common.Block
 		if i == 0 {
 			block = makeConfigBlock(i, prevHash, 0)
-			block.Header.DataHash = protoutil.BlockDataHash(block.Data)
+			block.Header.DataHash, _ = protoutil.BlockDataHash(block.Data)
```
Comment (tock-ibm) [inline orderer/common/follower/follower_chain_test.go]: why not ComputeBlockDataHash?

File: orderer/common/follower/follower_chain_test.go [hunk 26]
```diff
@@ -945,11 +945,11 @@ func (mbc *memoryBlockChain) fill(numBlocks uint64) {
...
 		} else {
 			block = protoutil.NewBlock(i, prevHash)
 			block.Data.Data = [][]byte{{uint8(i)}, {uint8(i)}}
-			block.Header.DataHash = protoutil.BlockDataHash(block.Data)
+			block.Header.DataHash, _ = protoutil.BlockDataHash(block.Data)
```
Comment (tock-ibm) [inline orderer/common/follower/follower_chain_test.go]: why not ComputeBlockDataHash?

File: orderer/common/follower/follower_chain_test.go [hunk 27]
```diff
@@ -963,7 +963,7 @@ func (mbc *memoryBlockChain) appendConfig(isMember uint8) {
 
 	h := uint64(len(mbc.chain))
 	configBlock := makeConfigBlock(h, protoutil.BlockHeaderHash(mbc.chain[h-1].Header), isMember)
-	configBlock.Header.DataHash = protoutil.BlockDataHash(configBlock.Data)
+	configBlock.Header.DataHash, _ = protoutil.BlockDataHash(configBlock.Data)
```
Comment (tock-ibm) [inline orderer/common/follower/follower_chain_test.go]: why not ComputeBlockDataHash?

File: orderer/common/follower/follower_chain_test.go [hunk 28]
```diff
@@ -998,7 +998,7 @@ func makeConfigBlock(num uint64, prevHash []byte, isMember uint8) *common.Block
 		),
 	}
 	block.Data.Data = append(block.Data.Data, protoutil.MarshalOrPanic(env))
-	block.Header.DataHash = protoutil.BlockDataHash(block.Data)
+	block.Header.DataHash, _ = protoutil.BlockDataHash(block.Data)
```
Comment (tock-ibm) [inline orderer/common/follower/follower_chain_test.go]: why not ComputeBlockDataHash?

File: orderer/common/multichannel/blockwriter_test.go [hunk 29]
```diff
@@ -62,8 +64,14 @@ func TestCreateBlock(t *testing.T) {
 		{Payload: []byte("some other bytes")},
 	})
 
+	dataHash := sha256.Sum256(bytes.Join(block.Data.Data, nil))
```
Comment (tock-ibm) [inline orderer/common/multichannel/blockwriter_test.go]: why not ComputeBlockDataHash?

File: orderer/consensus/etcdraft/blockcreator_test.go [hunk 30]
```diff
@@ -25,12 +25,14 @@ func TestCreateNextBlock(t *testing.T) {
 	}
 
 	second := bc.createNextBlock([]*cb.Envelope{{Payload: []byte("some other bytes")}})
+	hash := protoutil.ComputeBlockDataHash(second.Data)
```
Comment (tock-ibm) [inline orderer/consensus/etcdraft/blockcreator_test.go]: you can save a line here

File: orderer/consensus/etcdraft/blockcreator_test.go [hunk 31]
```diff
@@ -25,12 +25,14 @@ func TestCreateNextBlock(t *testing.T) {
...
 	require.Equal(t, first.Header.Number+1, second.Header.Number)
-	require.Equal(t, protoutil.BlockDataHash(second.Data), second.Header.DataHash)
+	require.Equal(t, hash, second.Header.DataHash)
 	require.Equal(t, protoutil.BlockHeaderHash(first.Header), second.Header.PreviousHash)
 
 	third := bc.createNextBlock([]*cb.Envelope{{Payload: []byte("some other bytes")}})
+	hash = protoutil.ComputeBlockDataHash(third.Data)
```
Comment (tock-ibm) [inline orderer/consensus/etcdraft/blockcreator_test.go]: you can save a line here

File: orderer/consensus/smartbft/assembler_test.go [hunk 32]
```diff
@@ -161,7 +161,7 @@ func makeConfigBlock(seq uint64) *common.Block {
 func proposalFromRequests(verificationSeq, seq, lastConfigSeq uint64, lastBlockHash, metadata []byte, requests ...[]byte) types.Proposal {
 	block := protoutil.NewBlock(seq, nil)
 	block.Data = &common.BlockData{Data: requests}
-	block.Header.DataHash = protoutil.BlockDataHash(block.Data)
+	block.Header.DataHash, _ = protoutil.BlockDataHash(block.Data)
```
Comment (tock-ibm) [inline orderer/consensus/smartbft/assembler_test.go]: why not ComputeBlockDataHash?

File: orderer/consensus/smartbft/verifier.go [hunk 33]
```diff
@@ -237,14 +237,14 @@ func verifyHashChain(block *cb.Block, prevHeaderHash string) error {
 		return errors.Errorf("previous header hash is %s but expected %s", thisHdrHashOfPrevHdr, prevHeaderHash)
 	}
 
-	if err := protoutil.VerifyTransactionsAreWellFormed(block); err != nil {
+	dataHashString := hex.EncodeToString(block.Header.DataHash)
```
Comment (tock-ibm) [inline orderer/consensus/smartbft/verifier.go]: it is probably better to compute that after the `if``

File: cmd/osnadmin/main_test.go (hunk 3 above)
Comment (yacovm) [reply cmd/osnadmin/main_test.go:823]: Why does it matter how I do it? the result is the same, and it's a test.

File: common/ledger/testutil/test_helper.go (hunk 4 above)
Comment (yacovm) [reply common/ledger/testutil/test_helper.go]: Why does it matter how I do it? the result is the same, and it's a test.

File: core/committer/txvalidator/v14/txvalidator_test.go (hunk 5 above)
Comment (yacovm) [reply core/committer/txvalidator/v14/txvalidator_test.go:280]: ok checking the error is a fair point. but why does it matter how I do it? the result is the same, and it's a test.

File: core/committer/txvalidator/v20/txvalidator_test.go (hunk 6 above)
Comment (yacovm) [reply core/committer/txvalidator/v20/txvalidator_test.go]: ok checking the error is a fair point. but why does it matter how I do it? the result is the same, and it's a test.

File: orderer/common/channelparticipation/validator.go (hunk 16 above)
Comment (yacovm) [reply orderer/common/channelparticipation/validator.go]: I removed it, because we only get config blocks so we cannot fail this check.

File: protoutil/blockutils_test.go (hunk 22 above)
Comment (yacovm) [reply protoutil/blockutils_test.go:614]: oh my god

File: orderer/common/channelparticipation/validator.go [hunk 34]
```diff
@@ -31,10 +29,6 @@ func ValidateJoinBlock(configBlock *cb.Block) (channelID string, err error) {
 		return "", errors.New("invalid block: does not have metadata")
 	}
 
-	if !bytes.Equal(protoutil.BlockDataHash(configBlock.Data), configBlock.Header.DataHash) {
-		return "", errors.New("invalid block: Header.DataHash is different from Hash(block.Data)")
-	}
-
```
Comment (tock-ibm) [inline orderer/common/channelparticipation/validator.go]: Why did you remove this test completely and not replace it with the new call? I agree that well formedness does not need to be checked after protoutil.IsConfigBlock but what abouts data hash vs the header.DataHash?

e.g.
```
if !bytes.Equal(protoutil.ComputeBlockDataHash(configBlock.Data), configBlock.Header.DataHash) {
		return "", errors.New("invalid block: Header.DataHash is different from Hash(block.Data)")
	}
```

File: orderer/common/channelparticipation/validator_test.go [hunk 35]
```diff
@@ -39,16 +39,6 @@ func TestValidateJoinBlock(t *testing.T) {
...
-		{
-			testName: "Invalid block data hash",
-			joinBlock: func() *cb.Block {
-				b := proto.Clone(validJoinBlock).(*cb.Block)
-				b.Header.DataHash = []byte("bogus")
-				return b
-			}(),
-			expectedChannelID: "",
-			expectedErr:       errors.New("invalid block: Header.DataHash is different from Hash(block.Data)"),
-		},
```
Comment (tock-ibm) [inline orderer/common/channelparticipation/validator_test.go]: Same question as above, I think this use case should be tested and protected against.

Comment (yacovm) [reply orderer/common/channelparticipation/validator_test.go]: done

File: orderer/common/channelparticipation/validator.go (hunk 34 above)
Comment (yacovm) [reply orderer/common/channelparticipation/validator.go]: You're right, my bad.

**Review by tock-ibm — APPROVED** [review]
